docs/Developers/Design/CRD_design.md

```diff
@@ -0,0 +1,438 @@
+  # Contains Secret w/ CA key & cert
+  glusterCA: ca-secret
+  # Georeplication
+  replication:  # (optional)
+    # Credentials for using this cluster as a target
+    credentials: secretRef
+    targets:
+      # Each target has a name that can be used in the StorageClass
+      - name: foo
+        # Addresses of node(s) in the peer cluster
+        address:
+          - 1.1.1.1
+          - my.dns.com
+        # Credentials for setting up session (ssh user & key)
+        credentials: SecretRef
+status:
+  # TBD operator state
+```
+
+All CRs live within the `operator.gluster.org` group and have version
+`v1alpha1`. The cluster `my-cluster`, above would be contained within the `gcs`
+namespace. All components of the Gluster cluster would be expected to exist
+within a single namespace. The `spec` field provides the main configuration
+options.
+
+The `clusterOptions` section are Gluster options (i.e., normally manipulated
+via ithe cli `gluster vol set`) that do not take a volume parameter.
+
+The `glusterCA` field holds a reference to a Kubernetes Secret containing the
+certificate authority `.key` and `.pem` files from which both client and server
+TLS keys can be generated. These will be used to automatically configure data
+encryption between the CSI driver and the Gluster bricks.
+
+The `replication` set of parameters define the geo-replication configuration
+for this cluster, optionally as both a source and target. If this cluster is to
+be used as a target for replication, the `replication.credentials` field must
+be supplied. This is a reference to a Secret that contains the inbound ssh user
+& key information. If this cluster is used as a source, the replication targets
+are presented as a list in `replication.targets`, providing a `name` for each
+remote cluster, the address(es) via `address`, and the ssh credentials via the
+`credentials` field.
+
+## Zone CR
+
+The `Zone` CR is used to define a failure domain. There would be at least one
+of these associated with a Gluster cluster. The purpose is to define how the
+operator can create new Gluster nodes within the failure domain and the rules
+for doing so. Additionally, it provides a name for the domain so that it can be
+referenced in the StorageClass to control volume placement and spreading.
+
+```yaml
+apiVersion: "operator.gluster.org/v1alpha1"
+kind: Zone
+metadata:
+  # Name for this zone
+  name: az1
+  # Zones are cluster-specific
+  cluster: my-cluster
+  # CRD is namespaced
+  namespace: gcs
+spec:
+  management:
+    # How the zone is managed
+    mode: automatic # (automatic | external | manual)
+    automatic:  # (iff mode==automatic)
+      # Minimum Gluster pods
+      minNodes: 1
+      # Maximum Gluster pods
+      maxNodes: 99
+      # Maximum raw storage capacity (South) in this zone
+      maxStorage: 100Ti
+      # Min unallocated capacity in the zone
+      freeStorageMin: 1Ti
+      # Max unallocated capacity in the zone
+      freeStorageMax: 3Ti
```

Comment:
Is operator going to calculate and update CR ? If yes, is it in a periodic manner and how ? Is based on the number of volumes in the cluster? if yes, does operator account node addition and deletion ?

Reply:
It would create and update the Node CRs, yes. My guess is that this would be done in response to alerts from agents (possibly within the operator, but not necessarily) monitoring # of volumes and total raw storage availability.

Reply:
If would update the `status:`, not these fields. These fields determine limits for autoscaling of the cluster. The operator would probe GD2 to ensure unallocated device capacity remains between Min & Max. It would then add/remove nodes if it gets out of bounds.